pkg/expr/graph.go

```diff
@@ -67,15 +72,34 @@ func (dp *DataPipeline) execute(c context.Context, now time.Time, s *Service) (m
 			dsNodes = append(dsNodes, node.(*DSNode))
 		}
 
-		if err := executeDSNodesGrouped(c, now, vars, s, dsNodes); err != nil {
-			return nil, err
-		}
+		executeDSNodesGrouped(c, now, vars, s, dsNodes)
 	}
 
 	for _, node := range *dp {
 		if groupByDSFlag && node.NodeType() == TypeDatasourceNode {
 			continue // already executed via executeDSNodesGrouped
 		}
+
+		// Don't execute nodes that have dependent nodes that have failed
+		var hasDepError bool
+		if cmd, ok := node.(*CMDNode); ok {
```

Comment:
> Can you please explain why we need to assert type here? Afaiu, as long as NeedsVars() returns some result, it is considered to be dependent, regardless its type.

Because `NeedsVars` is not on the Node interface, but will look at moving it there as that might make things cleaner (Perhaps not in this PR, will see...)

> Also, I wonder if we can move this into nodes, i.e. when we call Execute of a node that has dependency, the node itself checks that all related nodes' results are successful and return DependencyError if they are not.

I like it at this level _for now_ for the simplicity of it being in one place and not having to call it. That being said, I do expect this will change to be the way you describe it down the road. This is because I imagine there may be requests to do something to handle errors (like with math's nodata funcs for nodata).

Reply:
Added NeedsVars